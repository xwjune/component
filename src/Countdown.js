/**
 * 倒计时
 *
 * @param {number} count - 计数
 * @param {string} [type='hour'] - 类型【hour-小时，min-分钟】
 * @param {string } [prefix] - 文案前缀
 * @param {string } [suffix] - 文案后缀
 * @param {function } [onComplete] - 倒计时结束回调
 *
 * @author 小巷 <xwjune@163.com>
 * @date 2019/01/30
 */
import React from 'react';
import PropTypes from 'prop-types';

function zeroFill(data) {
  return data < 10 ? `0${data}` : data;
}

function secToMin(seconds) {
  return `${zeroFill(parseInt(seconds / 60, 10))}:${zeroFill(seconds % 60)}`;
}

export default class Countdown extends React.PureComponent {
  static propTypes = {
    count: PropTypes.number.isRequired,
    type: PropTypes.string,
    prefix: PropTypes.string,
    suffix: PropTypes.string,
    onComplete: PropTypes.func,
  };

  static defaultProps = {
    type: 'hour',
    prefix: '',
    suffix: '',
  };

  constructor(props) {
    super(props);
    this.state = {
      tip: this.renderTip(this.props.count),
    };
  }

  componentDidMount() {
    this.startTimer(this.props.count);
  }

  componentWillUnmount() {
    if (this.intervalId) {
      clearInterval(this.intervalId);
    }
  }

  startTimer = (count) => {
    this.intervalId = setInterval(() => {
      count--;
      if (count !== 0) {
        this.setState({
          tip: this.renderTip(count),
        });
      } else {
        this.setState({ tip: '' });
        clearInterval(this.intervalId);
        if (this.props.onComplete) {
          this.props.onComplete();
        }
      }
    }, 1000);
  };

  renderTip = (count) => {
    let result = '';
    switch (this.props.type) {
      case 'hour':
        result = `${secToMin(count)}`;
        break;
      case 'min':
        result = `${zeroFill(count)}`;
        break;
      // no default
    }
    return result;
  };

  render() {
    const { prefix, suffix } = this.props;
    const { tip } = this.state;

    return <span>{`${prefix}${tip}${suffix}`}</span>;
  }
}
