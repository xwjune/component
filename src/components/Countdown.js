/**
 * 倒计时
 *
 * @param {number} count - 计数
 * @param {string} [type='normal'] - 类型【normal-常规，hour-小时，minute-分钟】
 * @param {string } [prefix] - 文案前缀
 * @param {string } [suffix] - 文案后缀
 * @param {function } [onComplete] - 倒计时结束回调
 *
 * @date 2019/01/30
 */
import React from 'react';
import PropTypes from 'prop-types';

function zeroFill(data) {
  return data < 10 ? `0${data}` : data;
}

function secToMin(seconds) {
  return `${zeroFill(parseInt(seconds / 60, 10))}:${zeroFill(seconds % 60)}`;
}

export default class Countdown extends React.PureComponent {
  static propTypes = {
    count: PropTypes.number.isRequired,
    type: PropTypes.string,
    prefix: PropTypes.string,
    suffix: PropTypes.string,
    onComplete: PropTypes.func,
  };

  static defaultProps = {
    type: 'normal',
    prefix: '',
    suffix: '',
  };

  constructor(props) {
    super(props);
    this.state = {
      tip: this.renderTip(this.props.count),
    };
  }

  componentDidMount() {
    const { count } = this.props;
    if (count > 0) {
      this.startTimer(count);
    }
  }

  componentWillUnmount() {
    if (this.intervalId) {
      clearInterval(this.intervalId);
    }
  }

  startTimer = (count) => {
    this.intervalId = setInterval(() => {
      count--;
      if (count > 0) {
        this.setState({
          tip: this.renderTip(count),
        });
      } else {
        this.setState({ tip: '' });
        clearInterval(this.intervalId);
        if (this.props.onComplete) {
          this.props.onComplete();
        }
      }
    }, 1000);
  };

  renderTip = (count) => {
    if (count <= 0) {
      return '';
    }

    let result = '';
    switch (this.props.type) {
      case 'normal':
        result = count;
        break;
      case 'hour':
        result = `${secToMin(count)}`;
        break;
      case 'minute':
        result = `${zeroFill(count)}`;
        break;
      // no default
    }

    return result;
  };

  render() {
    const {
      prefix,
      suffix,
    } = this.props;
    const { tip } = this.state;

    return <span>{`${prefix}${tip}${suffix}`}</span>;
  }
}
