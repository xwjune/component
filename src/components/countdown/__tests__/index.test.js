import React from 'react';
import {
  shallow,
  mount,
} from 'enzyme';
import sinon from 'sinon';
import Countdown from '../index';

describe('Countdown', () => {
  beforeAll(() => {
    jest.useFakeTimers();
  });

  test('renders correctly', () => {
    const wrapper = shallow(<Countdown />);
    expect(wrapper.render()).toMatchSnapshot();

    const wrapper1 = shallow(
      <Countdown
        count={60}
        suffix="s"
      />
    );
    expect(wrapper1.render()).toMatchSnapshot();
  });

  test('count=0，组件不显示', () => {
    const wrapper = shallow(<Countdown />);
    expect(wrapper.find('span').exists()).toBeFalsy();
    expect(wrapper.isEmptyRender()).toBeTruthy();
    wrapper.unmount();
  });

  test('错误type', () => {
    const wrapper = shallow(
      <Countdown
        count={10}
        type="minute"
      />
    );
    expect(wrapper.find('span').text()).toBe('10');
  });

  test('倒计时', () => {
    const wrapper = shallow(
      <Countdown
        count={10}
        prefix="剩余"
        suffix="天"
      />
    );
    const wrapper1 = shallow(
      <Countdown
        count={60 * 60}
        type="hour"
      />
    );
    expect(wrapper.find('span').text()).toBe('剩余10天');
    expect(wrapper1.find('span').text()).toBe('60:00');
    jest.runTimersToTime(1000);
    expect(wrapper.find('span').text()).toBe('剩余9天');
    expect(wrapper1.find('span').text()).toBe('59:59');
  });

  test('1秒钟后回调onComplete', () => {
    const onComplete = jest.fn();
    const wrapper = shallow(
      <Countdown
        count={1}
        onComplete={onComplete}
      />
    );
    shallow(<Countdown count={1} />); // 未传入onComplete测试
    expect(wrapper.find('span').text()).toBe('1');
    expect(onComplete).not.toHaveBeenCalled();
    jest.runTimersToTime(1000);
    expect(wrapper.find('span').exists()).toBeFalsy();
    expect(wrapper.isEmptyRender()).toBeTruthy();
    expect(onComplete).toHaveBeenCalled();
  });

  test('componentWillUnmount', () => {
    sinon.spy(Countdown.prototype, 'componentWillUnmount');
    const wrapper = mount(<Countdown count={10} />);
    expect(Countdown.prototype.componentWillUnmount.calledOnce).toBeFalsy();
    wrapper.unmount();
    expect(Countdown.prototype.componentWillUnmount.calledOnce).toBeTruthy();
  });
});
