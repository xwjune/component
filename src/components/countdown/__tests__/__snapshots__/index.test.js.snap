// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Countdown renders correctly 1`] = `null`;

exports[`Countdown renders correctly 2`] = `
<span
  class=""
>
  60s
</span>
`;
